sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -57,6 +57,8 @@ abstract class Optimizer extends RuleExecutor[LogicalPlan] {
       ReplaceDistinctWithAggregate) ::
     Batch("Aggregate", FixedPoint(100),
       RemoveLiteralFromGroupExpressions) ::
+    Batch("Join", Once,
+      AddFilterOfNullForInnerJoin) ::
```

Comment:
We should make this rule idempotent instead of hacking it and making it run only once.  You are loosing the benefits of emergent optimizations with this implementation.

I would directly construct the filter in the left/right child, but only when its not already present in the `constraints` of the child.  This is the whole reason we added the ability to reason about what `constraints` are already present on a subtree.